lib/auth/recordingencryption/manager.go

```diff
@@ -0,0 +1,446 @@
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			return &types.EncryptionKeyPair{
+				PrivateKey:     privateKey,
+				PublicKey:      key.RecordingEncryptionPair.PublicKey,
+				PrivateKeyType: key.RecordingEncryptionPair.PrivateKeyType,
+			}, nil
+		}
+	}
+
+	return nil, trace.NotFound("no accessible decryption key found")
+}
+
+// GetAgeEncryptionKeys returns an iterator of AgeEncryptionKeys from a list of WrappedKeys. This is for use in
+// populating the EncryptionKeys field of SessionRecordingConfigStatus.
+func GetAgeEncryptionKeys(keys []*recordingencryptionv1.WrappedKey) iter.Seq[*types.AgeEncryptionKey] {
+	return func(yield func(*types.AgeEncryptionKey) bool) {
+		for _, key := range keys {
+			if !yield(&types.AgeEncryptionKey{
+				PublicKey: key.RecordingEncryptionPair.PublicKey,
+			}) {
+				return
+			}
+		}
+	}
+}
+
+// RecordingEncryptionResolver resolves RecordingEncryption state
+type RecordingEncryptionResolver interface {
+	ResolveRecordingEncryption(ctx context.Context) (*recordingencryptionv1.RecordingEncryption, error)
+}
+
+// WatchConfig captures required dependencies for building a RecordingEncryption watcher that
+// automatically resolves state.
+type WatchConfig struct {
+	Events        types.Events
+	Resolver      RecordingEncryptionResolver
+	ClusterConfig services.ClusterConfiguration
+	Logger        *slog.Logger
+	LockConfig    *backend.RunWhileLockedConfig
+}
+
+// A Watcher watches for changes to the RecordingEncryption resource and resolves the state for the calling
+// auth server.
+type Watcher struct {
+	events        types.Events
+	resolver      RecordingEncryptionResolver
+	clusterConfig services.ClusterConfiguration
+	logger        *slog.Logger
+	lockConfig    *backend.RunWhileLockedConfig
+}
+
+// NewWatcher returns a new Watcher.
+func NewWatcher(cfg WatchConfig) (*Watcher, error) {
+	switch {
+	case cfg.Events == nil:
+		return nil, trace.BadParameter("events is required")
+	case cfg.Resolver == nil:
+		return nil, trace.BadParameter("recording encryption resolver is required")
+	case cfg.ClusterConfig == nil:
+		return nil, trace.BadParameter("cluster config backend is required")
+	case cfg.LockConfig == nil:
+		return nil, trace.BadParameter("lock config is required")
+	}
+	if cfg.Logger == nil {
+		cfg.Logger = slog.With(teleport.ComponentKey, "encryption-watcher")
+	}
+
+	return &Watcher{
+		events:        cfg.Events,
+		resolver:      cfg.Resolver,
+		clusterConfig: cfg.ClusterConfig,
+		logger:        cfg.Logger,
+		lockConfig:    cfg.LockConfig,
+	}, nil
+}
+
+// Watch creates a watcher responsible for responding to changes in the RecordingEncryption resource.
+// This is how auth servers cooperate and ensure there are accessible wrapped keys for each unique keystore
+// configuration in a cluster.
+func (w *Watcher) Run(ctx context.Context) (err error) {
+	jitter := func() {
+		<-time.After(retryutils.SeventhJitter(time.Second * 5))
```

Comment:
I would change this function to also account for a Done ctx.
Currently, it will always wait for a least 5 seconds.

If we add a select statement for the `<-time.After` and `<-ctx.Done()` we can immediately exit.